Moving GetPropertyValue duplicate logic to the base class

ArtemTatarinov opened the pull request:
Fixes #6324
Continues PR #5745

## Proposed changes
- Moved the duplicate logic at the `GetPropertyValue` method to the base `AccessibleObject` class from its descendants
- Added unit tests for all cases of refactoring
- The base `AccessibleObject` class doesn't call the virtual `IsPatternSupported` method for the `IsInvokePatternAvailablePropertyId` case, but uses the large switch at the `IsInvokePatternAvailable` property instead, so we need to expand the cases of that switch for the new roles.

Also, I found out some questionable properties that I'd like to discuss:
1. The `false` value is always returned for all the `IsPasswordPropertyId` cases. Maybe we should move this property to the base class too?
2. There are 4 entries of the `FrameworkIdPropertyId` property, the value of `NativeMethods.WinFormFrameworkId` is returned in all the cases. Could it be moved to the base class too?
3. The `TopRowAccessibleObject` and `DataGridViewEditingPanelAccessibleObject` classes have the `HelpTextPropertyId` property case which returns `string.Empty`. Maybe we could delete it there, so the base class will return `Help ?? string.Empty`? By default `Help` is an empty string for both of these children classes.
4. The `AccessKeyPropertyId` property returns the `string.Empty` value for many cases at the `AccessibleObject` descendants. There was a PR #5705 where returning of the `KeyboardShortcut` was added to the `ControlAccessibleObject` class. What if we add the returning of the `KeyboardShortcut ?? string.Empty` to the base `AccessibleObject`? If its descendant has the KeyboardShortcut set, then it will be returned, that seems logical.
5. This one maybe is a bit tricky. For the `AutomationIdPropertyId` property 7 descendants of the `ControlAccessibleObject` class return the `Owner.Name` value, and 4 descendants of the `AccessibleObject` class return `AutomationId`. Maybe we should refactor them like that: base class will return `AutomationId` and `ControlAccessibleObject` will return `Owner.Name` at the overridden `GetPropertyValue` method?

## Customer Impact
- No

## Regression? 
- No

## Risk
- Minimal

## Test methodology <!-- How did you ensure quality? -->
- Manual (tested selected properties of some controls with the Inspect tool)
- Unit tests for all cases of refactoring
- CTI team

## Test environment(s) <!-- Remove any that don't apply -->
.NET SDK 7.0.0-alpha.1.21606.7
Microsoft Windows [Version 10.0.19043.1415]

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6429)

vladimir-krestov reviewed (commented):
Looks good!

ArtemTatarinov commented on src/System.Windows.Forms/src/System/Windows/Forms/CheckedListBox.CheckedListBoxItemAccessibleObject.cs:
```diff
@@ -44,7 +44,6 @@ public override void DoDefaultAction()
                 {
                     UiaCore.UIA.ControlTypePropertyId => UiaCore.UIA.CheckBoxControlTypeId,
-                    UiaCore.UIA.IsInvokePatternAvailablePropertyId => IsPatternSupported(UiaCore.UIA.InvokePatternId),
                     UiaCore.UIA.ValueValuePropertyId => Value,
```
- That's the only place in the codebase where the `ValueValuePropertyId` property is returned at the `GetPropertyValue` method. Should we move it to the base class anyway?
- The `LegacyIAccessibleNamePropertyId` property is used only one time too at the `ListViewGroupAccessibleObject` class, do we need to refactor it anyway?
- The `LegacyIAccessibleDefaultActionPropertyId` property is used twice: `DefaultAction` is returned at the `GridEntryAccessibleObject` class and `string.Empty` is returned at the `PropertyDescriptorGridEntryAccessibleObject` class. Should we return the `DefaultAction` value at the base class?
- The properties `LegacyIAccessibleChildIdPropertyId`, `LegacyIAccessibleValuePropertyId`, `LegacyIAccessibleDescriptionPropertyId`, `LegacyIAccessibleHelpPropertyId`, `LegacyIAccessibleKeyboardShortcutPropertyId`, `LegacyIAccessibleSelectionPropertyId` aren't used anywhere, I guess we could skip them?

vladimir-krestov replied:
In my opinion, we need to do that you described, 
1) for consistency
2) to avoid future issues
I think, everything should be uniform. @dreddy-work, @Tanya-Solyanik, your thoughts?

dreddy-work replied:
I agree.  we need to be consistent here.

Tanya-Solyanik replied:
Sounds reasonable, @ArtemTatarinov - please ensure that values of these properties do not regress in controls that were not implementing these properties in managed code.

ArtemTatarinov replied:
@Tanya-Solyanik I got two questions about the `ValueValuePropertyId` property, could you please look at them?

1. Am I getting this right, that if the `Value` property is implemented, then it should be returned at the `GetPropertyValue` method, and if it wasn't returned before refactoring, then it was an incorrect behavior, and refactoring will fix it? For example, the `DataGridViewTopRowAccessibleObject` class has `Value` override:
![image](https://user-images.githubusercontent.com/87859299/149273732-9cb11e2f-f817-4d46-9f00-45ac94e011bc.png)
But before refactoring `topRowAccessibleObject.GetPropertyValue(UiaCore.UIA.ValueValuePropertyId)` was returning `null`, because there was no `ValueValuePropertyId` case at the `GetPropertyValue` method of the `DataGridViewTopRowAccessibleObject` class. After refactoring it returns `"Top Row"` (its name by default), this behavior is new, but expected, because it's fixing incorrect behavior?
2. If the class doesn't have the `Value` property, the `null` value was returning - as default case at the `GetPropertyValue` method of the `AccessibleObject` class. After refactoring we'll be returning `string.Empty` value in some cases because of the original `Value` property implementation logic at the `AccessibleObject` class:
![image](https://user-images.githubusercontent.com/87859299/149275545-fda9d20d-c1eb-411e-a4e0-4cbb145d468a.png)
Is it okay?

Tanya-Solyanik replied:
> if the Value property is implemented, then it should be returned at the GetPropertyValue method for the ValueValuePropertyId

Sounds reasonable. 

>yet Inspect shows correct value

I think that MSAA Proxy is handling null in the before case.

>Is it okay to return string.Empty instead of null in these cases?

Good catch. I don't think we can replace null with an empty string...  Let'skeep this behavior- 
>If the class doesn't have the Value property, the null value was returning

ArtemTatarinov replied:
@Tanya-Solyanik I've pushed the latest changes and unit tests for them when `Value`/`LegacyIAccessibleDefaultAction`/`LegacyIAccessibleName` are returned as `null` for backward compatibility. But maybe we could return `string.Empty` after all since `GetPropertyValue` is an internal method and the Inspect/Accessibility Insights tools treat `null` and `string.Empty` values equally?

WPMGPRoSToTeMa commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxItemAccessibleObjectTests.cs:
```diff
@@ -219,12 +221,12 @@ public void ComboBoxItemAccessibleObject_GetPropertyValue_ScrollItemPattern_IsAv
                 ComboBoxItemAccessibleObject itemAccessibleObject = itemsCollection.GetComboBoxItemAccessibleObject(itemEntry);
 
-                Assert.True((bool)itemAccessibleObject.GetPropertyValue(UiaCore.UIA.IsScrollItemPatternAvailablePropertyId));
+                Assert.True((bool)itemAccessibleObject.GetPropertyValue((UiaCore.UIA)propertyId));
```
I think it would be better to check both patterns in a single test instead of making it a theory (a small optimization to not create a combo box twice).

ArtemTatarinov commented:
Refactored `NamePropertyId` and/or `HelpPropertyId` properties at the `ComboBoxChildEditUiaProvider`, `DataGridViewEditingPanelAccessibleObject`, `DropDownHolderAccessibleObject`, `ListViewColumnHeaderAccessibleObject`, `TopRowAccessibleObject` classes, and added unit tests for these cases if they weren't present already.

Tanya-Solyanik commented:
>The false value is always returned for all the IsPasswordPropertyId cases. Maybe we should move this property to the base class too?

Sounds reasonable. However, this does not have to be done in a single PR, please open an issue specifically for this property.

>There are 4 entries of the FrameworkIdPropertyId property, the value of NativeMethods.WinFormFrameworkId is returned in all the cases. Could it be moved to the base class too?

Yes, please. But in the next PR.

>The TopRowAccessibleObject and DataGridViewEditingPanelAccessibleObject classes have the HelpTextPropertyId property case which returns string.Empty. Maybe we could delete it there, so the base class will return Help ?? string.Empty? By default Help is an empty string for both of these children classes.

Sure, Vladimir also commented about Help property. If you do not have this change ready, then please put it into the next PR.

>The AccessKeyPropertyId property returns the string.Empty value for many cases at the AccessibleObject descendants. There was a PR Fix getting value for "ControlAccessibleObject.AccessKey" property #5705 where returning of the KeyboardShortcut was added to the ControlAccessibleObject class. What if we add the returning of the KeyboardShortcut ?? string.Empty to the base AccessibleObject? If its descendant has the KeyboardShortcut set, then it will be returned, that seems logical.

Sounds good, I would prefer it done in a separate PR and ensure that keyboard shortcut is not defined for non-controls.

>This one maybe is a bit tricky. For the AutomationIdPropertyId property 7 descendants of the ControlAccessibleObject class return the Owner.Name value, and 4 descendants of the AccessibleObject class return AutomationId. Maybe we should refactor them like that: base class will return AutomationId and ControlAccessibleObject will return Owner.Name at the overridden GetPropertyValue method?

This property is used in automation, it's really important to not regress it. If we make this change, I strongly recommend a dedicated PR in order to be able to roll it back if needed. Do you suggest moving `AutomationId` proeprty to the `AccessibleObject` class, making it virtual with the default implementation returning a `null` and implementing it as needed, i.e.  in the 4 classes derived from AO, keep the current implementations but mark them as overrides and in classes derived from the ControlAO, implement it as `Owner.Name`? This would be equivalent to the current behavior.

Tanya-Solyanik commented:
>And maybe we could move one step further and move returning of the Owner.Name to the ControlAccessibleObjectObject class itself to make it universal for the AutomationIdPropertyId property?

This makes sense, Name is a better default for pieces of UI that have Name property, but could be a breaking change for custom controls derived from our Control class that are navigates by UIS by ID. Please open a bug and flag this point for discussion in a team meeting. Personally, I think that a new release (7) can have a change like this one, but let's see what the team thinks.

>Сould we unite refactoring of the IsPasswordPropertyId, FrameworkIdPropertyId, HelpTextPropertyId and AccessKeyPropertyId properties into one issue (and future PR), these changes seem rather trivial? And refactoring of the AutomationIdPropertyId property will be in the second issue/PR which will be stand-alone.

Sounds good. This is up-to you. I only insist that `AutomationID` is in a dedicated PR, as a breaking change. For other cases, I did not want them in this PR, as it's growing too big.

ArtemTatarinov commented:
Created a new issue #6582 for the `IsPasswordPropertyId`, `FrameworkIdPropertyId`, `HelpTextPropertyId` and `AccessKeyPropertyId` properties.

Tanya-Solyanik commented:
@ArtemTatarinov  - please either skip or fix the failing tests

ArtemTatarinov commented:
> @ArtemTatarinov - please either skip or fix the failing tests

Looks like they were some temporary failures. I've run all tests in VS and after resolving merge conflicts all tests are good here too. Could you please look at this comment: https://github.com/dotnet/winforms/pull/6429#discussion_r795597495

Tanya-Solyanik commented:
@Olina-Zhang  - could you please test this change?

Olina-Zhang commented:
Tested this PR with no new issue found.
